core/src/main/java/org/elasticsearch/cluster/routing/IndexShardRoutingTable.java

```diff
@@ -262,6 +268,159 @@ public ShardIterator activeInitializingShardsIt(int seed) {
+        final int nodeCount = nodeIds.size();
+
+        final Map<String, Optional<ResponseCollectorService.ComputedNodeStats>> nodeStats = getNodeStats(nodeIds, collector);
+
+        // Retrieve all the nodes the shards exist on
+        final Map<String, Double> nodeRanks = rankNodes(nodeStats, nodeSearchCounts);
+
+        String minNode = null;
+        ResponseCollectorService.ComputedNodeStats minStats = null;
+        // calculate the "winning" node and its stats (for adjusting other nodes later)
+        for (Map.Entry<String, Optional<ResponseCollectorService.ComputedNodeStats>> entry : nodeStats.entrySet()) {
+            final String nodeId = entry.getKey();
+            final Optional<ResponseCollectorService.ComputedNodeStats> maybeNodeStats = entry.getValue();
+            if (maybeNodeStats.isPresent()) {
+                ResponseCollectorService.ComputedNodeStats stats = maybeNodeStats.get();
+                double rank = stats.rank(nodeSearchCounts.getOrDefault(nodeId, 1L));
+                if (minStats == null || rank < minStats.rank(nodeSearchCounts.getOrDefault(minStats.nodeId, 1L))) {
+                    minStats = stats;
+                    minNode = nodeId;
+                }
+            }
+        }
+
+        // sort all shards based on the shard rank
+        ArrayList<ShardRouting> sortedShards = new ArrayList<>(shards);
+        Collections.sort(sortedShards, new NodeRankComparator(nodeRanks));
+
+        // adjust the non-winner nodes' stats so they will get a chance to receive queries
+        adjustStats(collector, nodeStats, minNode, minStats);
+        if (minNode != null) {
+            // Increase the number of searches for the "winning" node by one.
+            // Note that this doesn't actually affect the "real" counts, instead
+            // it only affects the captured node search counts, which is
+            // captured once for each query in TransportSearchAction
+            nodeSearchCounts.compute(minNode, (id, conns) -> conns == null ? 1 : conns + 1);
+        }
+
+        return sortedShards;
+    }
+
+    private static class NodeRankComparator implements Comparator<ShardRouting> {
+        private final Map<String, Double> nodeRanks;
+
+        NodeRankComparator(Map<String, Double> nodeRanks) {
+            this.nodeRanks = nodeRanks;
+        }
+
+        @Override
+        public int compare(ShardRouting s1, ShardRouting s2) {
+            if (s1.currentNodeId().equals(s2.currentNodeId())) {
+                // these shards on the the same node
+                return 0;
+            }
+            Double shard1rank = nodeRanks.get(s1.currentNodeId());
+            Double shard2rank = nodeRanks.get(s2.currentNodeId());
+            if (shard1rank != null && shard2rank != null) {
+                return shard1rank.compareTo(shard2rank);
+            } else {
+                // One or both of the nodes don't have stats, treat them as equal
+                return 0;
+            }
```

Comment:
I'm wondering that the fact that the order is not transitive could confuse sorting. For instance if you have s1 and s2 so that s1 < s2 and s3 which is null then s1 and s2 are both equal to s3 but not equal with each other. Maybe we should make nulls always compare less than non nulls so that the order is total?

Reply:
I think it's okay to keep the contract of treating situations where both nodes do not have stats as equal, I also expect it to be a very very tiny margin of requests since `null` stats only occurs on a brand new node with 0 prior searches

Reply:
Actually javadocs of Arrays.sort that that an IllegalArgumentException may be thrown if the comparator [violates the Comparator contract](https://docs.oracle.com/javase/8/docs/api/java/util/Arrays.html#sort-T:A-java.util.Comparator-) and the Conparator javadocs say that it must implement a [total ordering](https://docs.oracle.com/javase/8/docs/api/java/util/Comparator.html) so I think it's important to make nulls compare consistently less than or greater than non-null values.

Reply:
Okay, it's not strictly nulls (since the nodes do exists, their `Optional`s are just empty, but I understand what you're saying. I'll change this to make missing values compare consistently less

Reply:
Doh they are nulls, I originally wrote it with `Optional` but it's different now, sorry!

Reply:
Okay I pushed 3d1dd2be7c8e59605e295f02f83c2187149eef94 for this